onmt/Models.py

```diff
@@ -425,6 +429,117 @@ def forward(self, src, tgt, lengths, dec_state=None):
             attns = None
         return out, attns, dec_state
 
+    @property
+    def lr(self):
+        return self.optimizer.param_groups[0]['lr']
+
+    @lr.setter
+    def lr(self, lr):
+        self.optimizer.param_groups[0]['lr'] = lr
+
+    def setup_optimizer(self, optimizer, initial_lr, max_grad_norm):
```

Comment:
Think this should go in the ModelController file. Maybe should be just called Controller now.

Reply:
I agree that this should be moved out of `NMTModel`. Since we would have a `Trainer.py`,  I think it makes more sense to be part of the duty of `Trainer`.    The static property of a model is in ` ModelConstructor.py`,  and the dynamic ones falls in the train phase and translate phase, which are `Trainer.py` and `Translator.py` , respectively.   As for the `Statistics` class,  I agree that it should be moved out `Loss.py` too,  but not in a `Log.py`.  The name is too vague, and it would guide people to think that it is logging utility or something like that.  I would like to put it in the `Trainer.py` too,  acting as a helper class for `Trainer`.

Further more,  this makes the most of the train logic self-containing in this file. Ideally,  after that, 
`train.py` would be acting as just a driver around the `Trainer`,  I think this could finish the https://github.com/OpenNMT/OpenNMT-py/issues/239.  

Comments ?  I will stop to start a new PR for this work first, after that I will come back for this PR.